biz.aQute.repository/src/aQute/bnd/repository/maven/provider/IndexFile.java

```diff
@@ -13,403 +11,527 @@
+			toBeDeleted.removeAll(toBeAdded);
+			toBeAdded.removeAll(older);
 
-		if (indexFile.lastModified() != lastModified && last + 10000 < System.currentTimeMillis()) {
-			loadIndexFile();
-			last = System.currentTimeMillis();
-			return true;
+			archives.keySet()
+				.removeAll(toBeDeleted);
+			return update(toBeAdded);
+		} else {
+			return promiseFactory.resolved(new BridgeRepository());
 		}
-		return descriptors.values()
-			.stream()
-			.filter(descriptor -> {
-				try {
-					if ((descriptor.promise != null) && descriptor.promise.isDone()
-						&& (descriptor.promise.getFailure() == null)) {
-						File f = descriptor.promise.getValue();
-						if ((f != null) && f.isFile() && (f.lastModified() != descriptor.lastModified)) {
-							updateDescriptor(descriptor, f);
-							return true;
-						}
-					}
-				} catch (InvocationTargetException | InterruptedException e) {
-					throw Exceptions.duck(e);
-				}
-				return false;
-			})
-			.count() > 0L;
-	}
 
-	private void loadIndexFile() throws Exception {
-		lastModified = indexFile.lastModified();
-		Set<Archive> toBeDeleted = new HashSet<>(descriptors.keySet());
+	}
 
-		if (indexFile.isFile()) {
-			lock.readLock()
-				.lock();
-			try (BufferedReader rdr = IO.reader(indexFile)) {
-				String line;
-				while ((line = rdr.readLine()) != null) {
-					line = Strings.trim(line);
-					if (line.startsWith("#") || line.isEmpty())
+	/*
+	 * Update the set of archives. This will add the given archives and then
+	 * create a new bridge promise. All newly added archives are scheduled to be
+	 * download and after downloaded are parsed. This method should only be
+	 * called when synchronized.
+	 */
+	private Promise<BridgeRepository> update(Collection<Archive> toAdd) throws Exception {
+
+		List<Promise<Boolean>> results = new ArrayList<>();
+		if (toAdd != null) {
+			for (Archive a : toAdd) {
+				put(a, info(a, "Loading ..."));
+
+				if (!a.isSnapshot()) {
+					File localFile = repo.toLocalFile(a);
+					if (localFile.isFile() && localFile.length() > 0) {
+						parseSingleOrMultiFile(a, localFile);
 						continue;
-
-					Archive a = Archive.valueOf(line);
-					if (a == null) {
-						reporter.error("MavenBndRepository: invalid entry %s in file %s", line, indexFile);
-					} else {
-						toBeDeleted.remove(a);
-						loadDescriptorAsync(a);
 					}
 				}
-			} finally {
-				lock.readLock()
-					.unlock();
-			}
 
-			this.descriptors.keySet()
-				.removeAll(toBeDeleted);
-			this.promises.keySet()
-				.removeAll(toBeDeleted);
+				Promise<Boolean> promise = repo.get(a)
+					.map(file -> {
+						if (file == null) {
+							return failed(a, "Not found");
+
+						} else {
+							parseSingleOrMultiFile(a, file);
+							return true;
+						}
+					})
+					.recover(p -> {
+						// TODO log
+						return failed(a, getMessage(p.getFailure()));
+					});
+
+				results.add(promise);
+			}
 		}
+
+		Collection<Resource> snapshot = archives.values();
```

Comment:
Since archives is a concurrent map, there is no point in the snapshot variable. Better to just pass `archives.values()` in the map lambda below.